### DISCUSSION TITLE
How to set a plane wave initial condition?

### DISCUSSION THREAD

**Discussion by ZhigangPu** [NanoComp/meep#2870] [Q&A] [answered]
Dear community,

I'm seeking assistance with setting an initial condition for a travelling plane wave in pymeep. While the tutorials extensively cover the usage of the Source class, I'm specifically interested in directly initializing the E and B fields within the simulation.

The documentation seems to focus primarily on source-based field generation, but I'm looking for a method to explicitly set the electromagnetic field values at the simulation's start. Is there a straightforward way to achieve this in pymeep?

I would greatly appreciate any guidance on how to directly manipulate the E and B fields to create an initial travelling plane wave condition. Perhaps there's a function or method I've overlooked?
Thank you in advance for your help and expertise.

Best regards,
Ayron

**Comment by stevengj** [comment, marked as answer]
> I'm looking for a method to explicitly set the electromagnetic field values at the simulation's start. Is there a straightforward way to achieve this in pymeep?

Yes, you can use the `initialize_field` method of the simulation class, as [described in the manual](https://meep.readthedocs.io/en/master/Python_User_Interface/#field-computations).

> I would greatly appreciate any guidance on how to directly manipulate the E and B fields to create an initial travelling plane wave condition.

Is your whole computational cell filled with a homogeneous material?  

* If so, why are you simulating it, since you have an analytical solution?

* If not, then a planewave by itself is not a solution of Maxwell's equations, so it's not clear to me what you mean by this "initial condition".   If you want an *incident* planewave, the cleanest way to do this is with a current source (as in many examples of the docs).

**Reply by ZhigangPu** [reply]
Thank you for your prompt response! Indeed, I do know the analytical solution of the traveling wave. My current focus is on observing how this simple plane wave interacts with an antenna within the computational domain.

I am actively developing new algorithms to solve Maxwell equations. The simple traveling wave is relatively straightforward to handle, as it doesn't require complex considerations for boundary conditions. However, the solutions of wave-antenna interactions remain unknown to me. This is precisely why I'm turning to the solutions provided by your program as a reference point. Your program's output serves as a valuable benchmark for validating my developing algorithms.

**Reply by stevengj** [reply]
> My current focus is on observing how this simple plane wave interacts with an antenna within the computational domain.

If you have an antenna within the computational domain, then a planewave everywhere in space does not solve Maxwell's equations by itself.   You need an *incident* planewave (created by a source), not an *initial* planewave (an initial condition filling all space).

**Reply by ZhigangPu** [reply]
Your approach is well-conceived. Let me refine my explanation:

The analytical solution I chosen for the Maxwell equations is simple for this simulation. Specifically, I am using

(𝐸𝑥,𝐸𝑦,𝐸𝑧,𝐵𝑥,𝐵𝑦,𝐵𝑧)=(0,0,𝑓(𝑥−𝑡),𝑓(𝑥−𝑡),0,0)

where $f(x)$ is a compact support function, i.e. [bump function](https://en.wikipedia.org/wiki/Bump_function).  

By using this analytical solution, I only need to specify the initial condition. The electromagnetic pulse, represented by the bump function, will then propagate according to this solution, allowing to observe its interaction with the antenna over time.

**Reply by stevengj** [reply]
Note that numerical dispersion (from the finite resolution) will still cause some artifacts (that is, this is no longer an exact solution), especially if you try to use a very short pulse (since very high frequencies are modeled inaccurately by a finite-difference scheme). But for a broad enough pulse you should see reasonably good convergence, I hope.

**Reply by ZhigangPu** [reply]
Thank you very much for your kind reminder. I will be sure to pay attention to this potential issue as I proceed with the simulation. Thanks again for your patience and thoroughness in this discussion.
